### PR TITLE
User and Authentication controller

### PR DESCRIPTION
**Pull request by Eidenn005**
User class, DTOs and service added
AuthenticationController added

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/exception/RegistrationException.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.example.onlinebookstore.exception;
+
+public class RegistrationException extends Exception {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/exception/RegistrationException.java:3]: Why this exception is checked?

File: src/main/java/com/example/onlinebookstore/repository/UserRepository.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User,Long> {
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/repository/UserRepository.java]:
```suggestion
public interface UserRepository extends JpaRepository<User, Long> {
```

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto save(
+            UserRegistrationRequestDto requestDto
+    ) throws RegistrationException {
+        if (userRepository.existsByEmail(requestDto.getEmail())) {
+            throw new RegistrationException("User already registered");
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java]:
```suggestion
            throw new RegistrationException("User with this emals: EMAIL is already registered");
```

File: src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        try {
+            var fieldValue = value.getClass().getDeclaredField(field);
+            var fieldMatchValue = value.getClass().getDeclaredField(fieldMatch);
+            fieldValue.setAccessible(true);
+            fieldMatchValue.setAccessible(true);
+
+            Object fieldVal = fieldValue.get(value);
+            Object fieldMatchVal = fieldMatchValue.get(value);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java]: Code duplication, logic of getting field could be moved to method

File: src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+            return fieldVal != null && fieldVal.equals(fieldMatchVal);
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java]: Is it possible to use Objects.equals?

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job, but some improvements required

File: src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java (hunk 4 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java]: Done

File: src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java (hunk 5 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/validation/impl/FieldMatchValidator.java]: Done

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java (hunk 3 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java]: done

File: src/main/java/com/example/onlinebookstore/repository/UserRepository.java (hunk 2 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/repository/UserRepository.java]: done

File: src/main/java/com/example/onlinebookstore/exception/RegistrationException.java (hunk 1 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/exception/RegistrationException.java:3]: Becouse 2 reasons:
1. App still need to be running, after exception occures
2. Its used to handle errors, coused by user input.

Leave unchanged.

**Review by andrii-hoienko — APPROVED** [review]
Nice

**Pull request merged**